src/mlpack/tests/kde_test.cpp

```diff
@@ -0,0 +1,821 @@
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::RTree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  kde.Evaluate(query, treeEstimations);
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < query.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(bfEstimations[i], treeEstimations[i], relError*100);
+}
+
+/**
+ * Test Standard Cover Tree dual-tree implementation results against brute
+ * force results.
+ */
+BOOST_AUTO_TEST_CASE(StandardCoverTreeGaussianKDETest)
+{
+  arma::mat reference = arma::randu(2, 500);
+  arma::mat query = arma::randu(2, 200);
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.3;
+  const double relError = 0.01;
+
+  // Brute force KDE
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::StandardCoverTree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  kde.Evaluate(query, treeEstimations);
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < query.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(bfEstimations[i], treeEstimations[i], relError*100);
+}
+
+/**
+ * Test duplicated value in reference matrix.
+ */
+BOOST_AUTO_TEST_CASE(DuplicatedReferenceSampleKDETest)
+{
+  arma::mat reference = arma::randu(2, 30);
+  arma::mat query = arma::randu(2, 10);
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.4;
+  const double relError = 0.05;
+
+  // Duplicate value
+  reference.col(2) = reference.col(3);
+
+  // Brute force KDE
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Dual-tree KDE
+  typedef KDTree<EuclideanDistance, kde::KDEStat, arma::mat> Tree;
+  std::vector<size_t> oldFromNewQueries, oldFromNewReferences;
+  Tree* queryTree = new Tree(query, oldFromNewQueries, 2);
+  Tree* referenceTree = new Tree(reference, oldFromNewReferences, 2);
+  KDE<GaussianKernel,
+      EuclideanDistance,
+      arma::mat,
+      KDTree>
+    kde(relError, 0.0, GaussianKernel(kernelBandwidth));
```

Comment:
Same here with double indentation, and a few other places in this file too it looks like.